node/src/runner.rs

```diff
@@ -105,17 +98,87 @@ pub async fn run(options: Options) -> Result<()> {
                 .with_context(|| format!("Failed to write config to the '{path:?}' file"))?;
             Ok(())
         }
-        Command::Run(ref run_options) => {
-            let node_config = NodeConfig::read(&options.config_path(), run_options)
-                .context("Failed to initialize config")?;
-            log::trace!("Starting with the following config\n: {node_config:#?}");
-            let manager = initialize(run_options.net, node_config).await?;
-            #[allow(clippy::unit_arg)]
-            Ok(manager.main().await)
+        Command::Mainnet(ref run_options) => {
+            let chain_config = common::chain::config::create_mainnet();
+            start(&options.config_path(), run_options, chain_config).await
+        }
+        Command::Regtest(ref regtest_options) => {
+            let chain_config = regtest_chain_config(&regtest_options.chain_config)?;
+            start(
+                &options.config_path(),
+                &regtest_options.run_options,
+                chain_config,
+            )
+            .await
         }
     }
 }
 
+async fn start(
+    config_path: &Path,
+    run_options: &RunOptions,
+    chain_config: ChainConfig,
+) -> Result<()> {
+    let node_config =
+        NodeConfig::read(config_path, run_options).context("Failed to initialize config")?;
+    log::trace!("Starting with the following config\n: {node_config:#?}");
+    let manager = initialize(chain_config, node_config).await?;
+    manager.main().await;
+    Ok(())
+}
+
+fn regtest_chain_config(options: &ChainConfigOptions) -> Result<ChainConfig> {
+    let ChainConfigOptions {
+        chain_address_prefix,
+        chain_blockreward_maturity,
+        chain_max_future_block_time_offset,
+        chain_version,
+        chain_target_block_spacing,
+        chain_coin_decimals,
+        chain_max_block_header_size,
+        chain_max_block_size_with_standard_txs,
+        chain_max_block_size_with_smart_contracts,
+    } = options;
+
+    let mut builder = ChainConfigBuilder::new(ChainType::Regtest);
+
+    macro_rules! update_builder {
+        ($field: ident) => {
+            paste! {
+                if let Some(val) = [<chain_ $field>] {
+                    builder = builder.$field(val.to_owned());
+                }
+            }
+        };
+        ($field: ident, $converter: stmt) => {
+            paste! {
+                if let Some(val) = [<chain_ $field>] {
+                    builder = builder.$field($converter(val.to_owned()));
+                }
+            }
+        };
```

Comment:
Could these two rules just delegate to the last (most general) one, with `std::convert::identity` used as the conversion function? Less code duplication at source level but it would expand into uglier code.

Reply:
The first branch can be delegated to the second one and it is a great idea, but I don't see how to delegate to the third branch: the types are different (`Result<T>` and `T`). Perhaps I'm missing something?

Reply:
Ok